nchelpers/__init__.py

```diff
@@ -1751,16 +1752,10 @@ def cmor_filename(self):
             extension='.nc', **self._cmor_type_filename_components()
         )
 
-    @property
+    @cached_property
     def unique_id(self):
-        """A unique id for this file, based on its CMOR filename"""
-        unique_id = cmor_type_filename(**self._cmor_type_filename_components())
-
-        dim_axes = set(self.dim_axes_from_names().values())
-        if not (dim_axes <= {'X', 'Y', 'Z', 'T'}):
-            unique_id += "_dim" + ''.join(sorted(dim_axes))
-
-        return unique_id.replace('+', '-')  # In original code, but why?
+        """A unique id for this file"""
+        return str(uuid.uuid4())
```

Comment:
This [generates a different random UUID every time it is executed](https://docs.python.org/3.6/library/uuid.html#uuid.uuid4). Which means that every time a program that uses `nchelpers` is run, a different UUID will be generated for any given dataset. 

But I think the way we use `unique_id` relies on it being stable across time. 

Possible problematic use case: Indexing. When a file is indexed in any given run, its `unique_id` will be generated (only once during that indexing run, but as a random value). Then when the same file is re-indexed in a later indexing run (it has moved location, say), its `unique_id` is generated, but it is by definition different than the first one. If the indexer (or anything else) tries to look up this file according to its `unique_id`, it will get incorrect results. Essentially, if anything (e.g., a modelmeta database) has a memory of a past `unique_id` for a file, it will be wrong the next time.

Reply:
Hmm, however, this argument also can be used against using metadata to generate the unique_id, since reindexing could in principle include reindexing a data file with slightly different metadata. But I think that case is implicitly excluded; the idea was that the metadata uniquely identified the file. We have an identity crisis here (pun, but serious too).

Reply:
Yeah, I was concerned about that too, so thank you for pointing this out.

Admittedly, I filed this PR more as provocation than as a finished product.

Maybe it would be better to change this from a (assumed to be static) property to a method (e.g. `create_unique_id()`), so users of it have to be intentional about when they want a new one (for example, during indexing) and when they want the existent one.

Reply:
That sounds more robust. 

Question is how to generate the unique id and, if it is not a deterministic computation based on the dataset, how to persist it. One solution is to make method `create_unique_id()` write its value to the metadata (attribute `unique_id`?) in the file, and for property `unique_id` to read it from there.

Reply:
Yeah, that's a great idea.